internal/prototest/yaml.go

```diff
@@ -0,0 +1,370 @@
+	nesting int
+}
+
+// message converts a Protobuf message into a [doc], which is used as an
+// intermediate processing stage to help compact make formatting decisions
+// (such as compressing nested messages).
+func (y *toYAML) message(m protoreflect.Message) *doc {
+	desc := m.Descriptor()
+	fs := desc.Fields()
+
+	d := new(doc)
+	for i := 0; i < fs.Len(); i++ {
+		f := fs.Get(i)
+
+		has := m.Has(f)
+		if y.EmitZeros && !has && !f.HasPresence() {
+			has = true
+		}
+		if !has {
+			continue
+		}
+
+		d.push(
+			f.Name(),
+			y.value(m.Get(f), f),
+		)
+	}
+	return d
+}
+
+// value converts a Protobuf value into a value that can be placed into a
+// [doc].
+func (y *toYAML) value(v protoreflect.Value, f protoreflect.FieldDescriptor) any {
+	switch v := v.Interface().(type) {
+	case protoreflect.Message:
+		return y.message(v)
+
+	case protoreflect.List:
+		d := new(doc)
+		for i := 0; i < v.Len(); i++ {
+			d.push(nil, y.value(v.Get(i), f))
+		}
+		return d
+
+	case protoreflect.Map:
+		d := new(doc)
+		d.needsSort = true
+		v.Range(func(k protoreflect.MapKey, v protoreflect.Value) bool {
+			d.push(
+				y.value(k.Value(), f.MapKey()),
+				y.value(v, f.MapValue()),
+			)
+			return true
+		})
+		return d
+
+	case protoreflect.EnumNumber:
+		enum := f.Enum()
+		if value := enum.Values().ByNumber(v); value != nil {
+			return value.Name()
+		} else {
+			return int32(v)
+		}
+
+	case []byte:
+		return string(v)
+
+	default:
+		return v
+	}
+}
+
+// write writes a value returned by [toYAML.value] into the internal output
+// buffer.
+func (y *toYAML) write(v any) {
+	switch v := v.(type) {
+	case int32, int64, uint32, uint64, float32, float64, protoreflect.Name:
+		fmt.Fprint(&y.out, v)
+	case string:
+		fmt.Fprintf(&y.out, "%q", v)
+	case *doc:
+		isOneLine := func(v any) bool {
+			maxWidth := y.MaxWidth
+			if maxWidth == 0 {
+				maxWidth = 80
+			}
+			maxWidth -= y.nesting * 2
+
+			doc, ok := v.(*doc)
+			return !ok || doc.width < maxWidth
+		}
+
+		if isOneLine(v) {
+			switch {
+			case v.isArray:
+				y.out.WriteString("[")
+				for i, pair := range v.pairs {
+					if i > 0 {
+						y.out.WriteString(", ")
+					}
+					y.write(pair[1])
+				}
+				y.out.WriteString("]")
+
+			case len(v.pairs) == 0:
+				y.out.WriteString("{}")
+
+			default:
+				// Special case: if we are a list element, and there is only
+				// one entry, print it directly.
+				if len(v.pairs) == 1 && strings.HasSuffix(y.out.String(), "- ") {
+					y.write(v.pairs[0][0])
+					y.out.WriteString(": ")
+					y.write(v.pairs[0][1])
+					return
+				}
+
+				y.out.WriteString("{ ")
+				for i, pair := range v.pairs {
+					if i > 0 {
+						y.out.WriteString(", ")
+					}
+					y.write(pair[0])
+					y.out.WriteString(": ")
+					y.write(pair[1])
+				}
+				y.out.WriteString(" }")
+			}
+			return
+		}
+
+		for _, pair := range v.pairs {
+			oneLine := isOneLine(pair[1])
+			y.indent()
+
+			if pair[0] == nil {
+				y.out.WriteString("- ")
```

Comment:
Do we not need to also increment `y.nesting` to account for this, to avoid subsequent lines lining up under the `- ` instead of under the prior element?
```yaml
list:
  - foo: bar
  baz: bot  # oops
  - foo: bazzle
```

Or we could be a little more careful when writing the indentation to omit the last two space and effectively replace with the "- ":
```yaml
list:
- foo: bar
  baz: bot
- foo: bazzle
```

Reply:
This is already handled by the y.nesting++ line below.

Reply:
But the condition for that increment below doesn't consider whether the bullet prefix is used. Wouldn't that result in output like the first example above? In that first example above, elements in an array would need _two_ increments (i.e. the "baz" property needs to be indented four spaces, not just two, because it is in an array).

Reply:
Yes, but the first element will not be indented, because `y.indent()` is called before we append the `-` and only after that do we increment the indentation level.

Unfortunately there's no tests for this in this PR because the "tests" are the generated test data in the parser PR. But trust me, it gets this right :P